Troubleshooting for Survey Library (#4992)
surveyjs/survey-library: docs/Troubleshooing.md

    > NOTE: Instead of a JSON schema, you can use methods to build a survey model. Refer to the following documentation article for details: [Create or Change a Survey Model Dynamically](https://surveyjs.io/form-library/documentation/design-survey-create-a-simple-survey#create-or-change-a-survey-model-dynamically).

* **Survey does not contain visible elements**         
Survey elements become invisible when their [visible](https://surveyjs.io/form-library/documentation/question#visible) property is set to `false` or their [visibleIf](https://surveyjs.io/form-library/documentation/question#visibleIf) expression evaluates to `false`. Ensure that the `visible` property is enabled and the `visibleIf` expressions evaluate to `true`.

> * **Survey uses an unregistered custom question component**       
> 

Reviewer: ```suggestion
* **Survey uses an unregistered custom question component**        
```

# Troubleshooting

This help topic describes how to troubleshoot problems that may occur when you use the SurveyJS library in your application.

## There is no visible page or question in the survey.

> **Issue**: The survey appears empty and the following message is displayed: "There is no visible page or question in the survey".

Author: OK, created a ticket: https://github.com/surveyjs/survey-library/issues/5057

> **Solution**:
> You may encounter this issue in the following cases.

Reviewer: ```suggestion
**Solution**:

You may encounter this issue in the following cases:
```

> * A survey JSON is bad-formatted.

Reviewer: ```suggestion
* **Survey JSON is invalid**        
```

> A survey model can be built from JSON. A survey JSON schema can be declared in your appplication, or imported from a file. 
> 
> Validate JSON and ensure it is well-formatted. Also, ensure that the JSON schema follows the [survey structure](https://surveyjs.io/form-library/documentation/design-survey-create-a-simple-survey#survey-structure) and declares the `elements` node at a minimum. For more information, refer to: [Define a Static Survey Model in JSON](https://surveyjs.io/form-library/documentation/design-survey-create-a-simple-survey#define-a-static-survey-model-in-json).
> 
> > NOTE: It is possible to programmatically build a survey model using the [addNewPage](https://surveyjs.io/form-library/documentation/surveymodel#addNewPage), [addNewPanel](https://surveyjs.io/form-library/documentation/pagemodel#addNewPanel), and [addNewQuestion](https://surveyjs.io/form-library/documentation/pagemodel#addNewQuestion) functions. For more details, refer to [Create or Change a Survey Model Dynamically](https://surveyjs.io/form-library/documentation/design-survey-create-a-simple-survey#create-or-change-a-survey-model-dynamically).

Reviewer: ```suggestion
When you declare a survey JSON schema or load it from a file or a web server, ensure that it does not contain syntax errors (unclosed brackets or quotation marks, missing commas). Your JSON schema should also follow the [survey structure](https://surveyjs.io/form-library/documentation/design-survey-create-a-simple-survey#survey-structure) and declare at least the `elements` property at the first level. Refer to the following help topic for more information: [Define a Static Survey Model in JSON](https://surveyjs.io/form-library/documentation/design-survey-create-a-simple-survey#define-a-static-survey-model-in-json).

    > NOTE: Instead of a JSON schema, you can use methods to build a survey model. Refer to the following documentation article for details: [Create or Change a Survey Model Dynamically](https://surveyjs.io/form-library/documentation/design-survey-create-a-simple-survey#create-or-change-a-survey-model-dynamically).
```

> * A survey does not contain visible elements.

Reviewer: ```suggestion
* **Survey does not contain visible elements**         
```

> Survey elements may have the [visible](https://surveyjs.io/form-library/documentation/question#visible) property set to `false`. Or, their [visibleIf](https://surveyjs.io/form-library/documentation/question#visibleIf) expression evaluates to `false`.
> 
> Ensure the [visible](https://surveyjs.io/form-library/documentation/question#visible) option for survey elements is enabled or the [visibleIf](https://surveyjs.io/form-library/documentation/question#visibleIf) expression evaluates to `true`.

Reviewer: ```suggestion
Survey elements become invisible when their [visible](https://surveyjs.io/form-library/documentation/question#visible) property is set to `false` or their [visibleIf](https://surveyjs.io/form-library/documentation/question#visibleIf) expression evaluates to `false`. Ensure that the `visible` property is enabled and the `visibleIf` expressions evaluate to `true`.
```

> * A survey uses a custom question component which is not registered on a page.

Reviewer: ```suggestion
* **Survey uses an unregistered custom question component**       
```

> If you use a custom question component, ensure that it is registered in a survey's `ComponentCollection`. For more information, refer to [Create Specialized Question Types](https://surveyjs.io/survey-creator/documentation/create-specialized-question-types).
> 
>  
> 

Reviewer: ```suggestion
If you implement a custom question component, add it to the survey's `ComponentCollection`. For more information, refer to the following help topic: [Create Specialized Question Types](https://surveyjs.io/survey-creator/documentation/create-specialized-question-types).
```
